pkg/rpcserver/runner.go

```diff
@@ -52,14 +57,81 @@ type runnerStats struct {
 	statNoExecDuration     *stats.Val
 }
 
-func (runner *Runner) connectionLoop() error {
+type handshakeConfig struct {
+	VMLess     bool
+	Timeouts   targets.Timeouts
+	LeakFrames []string
+	RaceFrames []string
+	Files      []string
+	Globs      []string
+	Features   flatrpc.Feature
+
+	// Callback() is called in the middle of the handshake process.
+	// The return arguments are the coverage filter and the (possible) error.
+	Callback func(*flatrpc.InfoRequestRawT) (handshakeResult, error)
+}
+
+type handshakeResult struct {
+	CovFilter     []uint64
+	MachineInfo   []byte
+	Canonicalizer *cover.CanonicalizerInstance
+}
+
+func (runner *Runner) Handshake(conn *flatrpc.Conn, cfg *handshakeConfig) error {
+	connectReply := &flatrpc.ConnectReply{
+		Debug:            runner.debug,
+		Cover:            runner.cover,
+		CoverEdges:       runner.coverEdges,
+		Kernel64Bit:      runner.sysTarget.PtrSize == 8,
+		Procs:            int32(runner.procs),
+		Slowdown:         int32(cfg.Timeouts.Slowdown),
+		SyscallTimeoutMs: int32(cfg.Timeouts.Syscall / time.Millisecond),
+		ProgramTimeoutMs: int32(cfg.Timeouts.Program / time.Millisecond),
+		LeakFrames:       cfg.LeakFrames,
+		RaceFrames:       cfg.RaceFrames,
+		Files:            cfg.Files,
+		Globs:            cfg.Globs,
+		Features:         cfg.Features,
+	}
+	if err := flatrpc.Send(conn, connectReply); err != nil {
+		return err
+	}
+	infoReq, err := flatrpc.Recv[*flatrpc.InfoRequestRaw](conn)
+	if err != nil {
+		return err
+	}
+	ret, err := cfg.Callback(infoReq)
+	if err != nil {
+		return err
+	}
+	infoReply := &flatrpc.InfoReply{
+		CoverFilter: ret.CovFilter,
+	}
+	if err := flatrpc.Send(conn, infoReply); err != nil {
+		return err
+	}
+	runner.mu.Lock()
+	runner.conn = conn
+	runner.machineInfo = ret.MachineInfo
+	runner.canonicalizer = ret.Canonicalizer
+	runner.mu.Unlock()
+	return nil
+}
+
+func (runner *Runner) ConnectionLoop() error {
+	runner.mu.Lock()
```

Comment:
These lock is not necessary, right? Handshake must be called before ConnectionLoop.
I carefully wrote runner loop to not take any locks, but we have a bunch, but all of them also look unnecessary.
Let's remove them at least from runner loop called functions.

Reply:
As of now, `DistributeSignalDelta()` may well be called concurrently to everything else, which in turn at least invokes the `Alive()` method.

Do we want to access the same set of fields (e.g. `conn`) in some cases under a mutex and in some cases without it?

Reply:
Yes, Alive still needs the mutex.

> Do we want to access the same set of fields (e.g. conn) in some cases under a mutex and in some cases without it?

Yes, I would say so.
Sprinkling mutexes everywhere for no reason just confuses readers (and makes things slower).
When I see a mutex I am trying to understand why it is there, and what concurrent mutations can happen concurrnetly.
In some methods we just read runner.conn under mutex and then use it w/o checks. This does not make sense. If it's changing concurrently between nil/non-nil, then we also need some checks. If it's not changing, then we don't need the mutex either.
Just-in-case mutexes also disable race detector from producing useful reports about unexpected concurrency.

Reply:
Done